### PR TITLE
Use SPIRE to manage cert for CRD k8s-workload-registrar webhook

### PR DESCRIPTION
**Pull request by faisal-memon**
Signed-off-by: Faisal Memon <f.memon@f5.com>

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
k8s-workload-registrar CRD

**Description of change**
Simplifies deployment of the CRD k8s-workload-registrar by using SPIRE certificates for the validating webhook, instead of user-provided. 

- Leverages https://github.com/spiffe/spire/pull/2022 to have SPIRE rotate the caBundle.
- Upon startup creates an entry for the webhook, and deletes the entry on exit. (mode-crd/webhook/webhook_entry.go)
- Uses a watcher, similar to this [example](https://github.com/spiffe/go-spiffe/blob/main/v2/examples/spiffe-watcher/main.go)  to maintain the cert/key in the k8s-workload-registrar container. (mode-crd/webhook/webhook_svid_watcher.go)
- Backwards compatible

Includes a quick start tutorial (mode-crd/README.md) to make it easier to get started.

### CODE REVIEW

File: support/k8s/k8s-workload-registrar/mode-crd/config/spire-server-registrar.yaml
Comment (faisal-memon) [inline support/k8s/k8s-workload-registrar/mode-crd/config/spire-server-registrar.yaml:240]: The prestop hook is used to ensure the entry created for the webhook is cleaned up. It tell the spire server to wait for the k8s-workload-registrar to exit first.

**Comment by azdagron** [comment]
Also, looks like we're missing DCO signoff on your latest commit?

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_entry.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_entry.go:212]: It will if the annotation/label entry has the same spiffe id. The entry won't be deleted unless the pod its attached to has been deleted. It checks to make sure no pod with a uid matching the selector uid exists before deleting it. In that case it would be deleted anyways, so it won't cause a problem. https://github.com/spiffe/spire/pull/2321/files#diff-9355d9bfd9c454f8c9630d5a02c2c50dcf14fd6c5f23b1175021b5a009ac3100R124-R139

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_entry.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_entry.go]: thanks, this is fixed.

File: support/k8s/k8s-workload-registrar/mode-crd/controllers/utils.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/controllers/utils.go]: Yea makes sense to have it go through the normal error handling path.

File: support/k8s/k8s-workload-registrar/config_crd.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/config_crd.go:154]: The agent socket path is not a required configurable. It is currently only used when the registrar is not colocated with SPIRE server and needs to obtain an admin SVID. Up until now there has been no prerequisite that the registrar be colocated with an agent. From what I understand, this change requires that the registrar go through an agent to obtain an SVID for hosting the webhook? I wonder if the mint API would be appropriate to use instead. Paging @evan2645.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_entry.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_entry.go:39]: Stashing context's away in structs in usually frowned upon. It's preferred to pass the context in as the first parameter to the function that uses it. That has the added benefit of self documentation as to which methods may be long running or may return a failure due to cancellation.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go:116]: dumpSVID?

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go:87]: Also need to select on s.c.Ctx here in case the outer context is canceled before the three minutes are up.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go:144]:
```suggestion
	var svid *x509svid.SVID
```

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go:153]: We can short cirtcuit the search once we've found the right candidate...
```suggestion
		for _, candidate := range c.SVIDs {
			if candidate.ID.TrustDomain().String() == w.s.c.SpiffeID.GetTrustDomain() &&
				candidate.ID.Path() == w.s.c.SpiffeID.GetPath() {
				svid = candidate
				break
			}
		}
```

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go:47]: nit: this is a fairly inconvenient type; i'd suggest using `spiffeid.ID` where possible except when interacting with the API. For example, in this watcher, having this be a spiffeid.ID would allow a simple `==` comparison when finding the candidate SVID below.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_watcher.go:167]: Why don't we want to see PermissionDenied? I understand there is probably a time period when this is first coming up where the entry has yet to be created/propogated but if something happened to the entry later, wouldn't we want to know about it?

**Review by azdagron — COMMENTED** [review]
I'm still digesting these changes... but here are some comments in the meantime.

File: support/k8s/k8s-workload-registrar/config_crd.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/config_crd.go:154]: Yes, we use the agent to get the SVID for the webhook. Is there a way to keep the SVIDs up to date with the mint API as the agent does?

Comment (azdagron) [reply support/k8s/k8s-workload-registrar/config_crd.go:154]: Not without manually tracking the expiration and requesting a new SVID before the old one expires. Since this would involve a single identity, it doesn't seem too onerous to do that. Curious what @evan2645 thinks.

**Comment by azdagron** [comment]
The maintainers discussed this PR today and came to consensus that we should do the approach discussed during the contributor sync, namely we should rely on the registrar's admin capabilities to mint an SVID directly using the (SVID.MintX509SVID](https://github.com/spiffe/spire-api-sdk/blob/7ff3eb0759ce242a38fd02aa368b3d7a33ac7316/proto/spire/api/server/svid/v1/svid.proto#L15) RPC and manage it's own rotation schedule. It is probably appropriate to adopt the naive strategy implemented by the agent, namely to attempt renewal when the SVID lifetime is half over (with some backoff on failure).

This approach should drastically simplify the PR since there will no longer be a need to register and unregister the webhook entry.

How does that feel to you @faisal-memon ?

**Comment by faisal-memon** [comment]
> The maintainers discussed this PR today and came to consensus that we should do the approach discussed during the contributor sync, namely we should rely on the registrar's admin capabilities to mint an SVID directly using the (SVID.MintX509SVID](https://github.com/spiffe/spire-api-sdk/blob/7ff3eb0759ce242a38fd02aa368b3d7a33ac7316/proto/spire/api/server/svid/v1/svid.proto#L15) RPC and manage it's own rotation schedule. It is probably appropriate to adopt the naive strategy implemented by the agent, namely to attempt renewal when the SVID lifetime is half over (with some backoff on failure).
> 
> This approach should drastically simplify the PR since there will no longer be a need to register and unregister the webhook entry.
> 
> How does that feel to you @faisal-memon ?

Makes sense to me, itll make the implementation alot simpler. Where in the agent code is the rotation logic?

**Comment by azdagron** [comment]
It uses this helper: https://github.com/spiffe/spire/blob/27c9cf10daf5811ba2dd9b7173cfaebd375133a5/pkg/common/rotationutil/rotationutil.go#L12

**Comment by faisal-memon** [comment]
@azdagron This is now updated to use MintX509SVID

File: support/k8s/k8s-workload-registrar/config_crd.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/config_crd.go:194]: nit: this would be a little more robust if we did not do the stat and instead handled the proper error condition returned from `unix.Access`:
```
func (c *CRDMode) certDirExistsAndReadOnly() (bool, error) {
	err := unix.Access(c.WebhookCertDir, unix.W_OK)
	switch {
	case err == nil, errors.Is(err, unix.ENOENT):
		return false, nil
	case errors.Is(err, unix.EROFS):
		return true, nil
	default:
		return false, err
	}
}
```

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:73]:
```suggestion
		return nil, fmt.Errorf("unable to generate SPIFFE ID: %w", err)
```

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:111]: This is probably good enough for this purpose but in general this timestamp should be obtained from the certificate. It shouldn't be a problem here, but I thought I'd point it out as a red flag.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:166]: These files (both cert and key) aren't being closed. However, in this case, you can just use `os.WriteFile`...
```suggestion
	if err := os.WriteFile(certsFileName, svidPEM.Bytes(), certsFileMode); err != nil {
		return err
	}
```

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:177]:
```suggestion
	keyFileName := path.Join(e.c.WebhookCertDir, keyFileName)
	if err := os.WriteFile(keyFileName, keyPEM, keyFileMode)
		return err
	}
```

File: support/k8s/k8s-workload-registrar/config_crd.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/config_crd.go:131]: It seems problematic to take down the registrar on a transient failure to mint (which is totally a possibility depending on the keymanager in use). I'd recommend not cycling the registrar until the webhook certificate is in immediate danger of becoming expired (or has expired).

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: See my other comment. If we change this code to no longer cycle the registrar on failure, then we'll also need to implement a backoff strategy so we don't pummel the server with requests.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:111]: yes, youre right. Do you know offhand if the leaf certificate is the first or the last in the chain in the mint svid response?

Comment (azdagron) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:111]: leaf is first

File: support/k8s/k8s-workload-registrar/config_crd.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/config_crd.go:194]: Thanks for the suggestion. This change is in.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: Yes good point. How about using `retry.OnError()`? https://pkg.go.dev/k8s.io/client-go/util/retry#OnError

Comment (azdagron) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: That's probably ok? What do you plan on doing inside of the "retriable" callback?

Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: These lines of code: https://github.com/spiffe/spire/pull/2321/files#diff-ff4a359cdb6e273cd555b8aa198c0d14bed25a991da06708c8e7453c4642dc8bR104-R115

Comment (azdagron) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: That seems like what you'd put in `fn`. `retriable` classifies the given error as retriable or not. I was just curious if you were going to classify all errors as retriable or what.

Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: Is there an error we don't want to retry on?

Comment (azdagron) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:126]: I don't think so. One thing we may consider doing is returning false if the current SVID has expired, which will cause 
OnError to return an error, which we can then use to tear down the registrar (since the webhook is now presumably non-functional).

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go]: this change is in.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:146]: This change is in.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:42]: Shouldn't we have the executable bit on the directory?

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:85]: This is ok for now, but there is some shared state on the SVID type that is could end up being accessed by multiple goroutines based solely on the exported functions from the SVID type. I don't necessarily think we need to fix it now, but thought I'd put it on your radar in case this code is changed in future refactor.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:1]: We haven't traditionally added the copyright clause to each source file and for the sake of consistency should probably exclude it here. Whether or not adding the clause is a good idea is topic of debate 😆

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_test.go
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid_test.go:1]: Same here.

**Review by azdagron — COMMENTED** [review]
I think this is pretty good, @faisal-memon . I left a few minor comments, but I think we are super close.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:42]: I don't think any of them need the executable bit do they?

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (faisal-memon) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:85]: Yes good point. We should also make sure only one instance of the SVIDRotator is running.

File: support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go
Comment (azdagron) [reply support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go:42]: The directory itself needs the executable bit or you won't have sufficient permissions to open files inside of it.

**Comment by azdagron** [comment]
Hey faisal, I think we are good here. Mind resolving the merge conflicts?

**Comment by faisal-memon** [comment]
> Hey faisal, I think we are good here. Mind resolving the merge conflicts?

Yes. It it ok to rebase or is there another way to resolve?

**Comment by azdagron** [comment]
Whatever is easiest for you at this point since the review is all but wrapped up.

**Comment by azdagron** [comment]
CI/CD has a few issues after the rebase.

**Comment by faisal-memon** [comment]
> CI/CD has a few issues after the rebase.

Yep, working on those right now. Should be fixed soon.

**Review by azdagron — APPROVED** [review]

**Pull request merged**
